zipkin/src/main/java/com/linecorp/armeria/common/tracing/AbstractTracingBuilder.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.tracing;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+import javax.annotation.Nullable;
+
+import brave.Clock;
+import brave.ErrorParser;
+import brave.Tracing;
+import brave.Tracing.Builder;
+import brave.handler.FinishedSpanHandler;
+import brave.propagation.CurrentTraceContext;
+import brave.propagation.Propagation;
+import brave.propagation.Propagation.Factory;
+import brave.sampler.Sampler;
+import zipkin2.Span;
+import zipkin2.reporter.Reporter;
+
+/**
+ * Builds a new {@link AbstractTracingBuilder} or its decorator function.
+ *
+ * @param <B> the self type
+ */
+public abstract class AbstractTracingBuilder<B extends AbstractTracingBuilder<B>> {
+
+    @Nullable
+    private Builder tracingBuilder;
+
+    @Nullable
+    private Tracing tracing;
+
+    private boolean isCurrentTraceContextSet;
+
+    @SuppressWarnings("unchecked")
+    private B self() {
+        return (B) this;
+    }
+
+    /**
+     * Sets the {@link Tracing}.
+     *
+     * @throws IllegalStateException if other setters for building a new {@link Tracing} are already invoked
+     */
+    public B tracing(Tracing tracing) {
+        checkState(tracingBuilder == null, "tracingBuilder is already set.");
```

Comment:
Could we rephrase the exception message from user's standpoint? I don't think a user will know what this really means.